### ISSUE TITLE
Update to xunit v3

### ISSUE THREAD

**Issue by pengweiqhca** [AArnott/Xunit.StaFact#103]
- https://xunit.net/docs/getting-started/v3/whats-new
- https://xunit.net/docs/getting-started/v3/migration
- https://xunit.net/docs/getting-started/v3/cmdline

**Comment by AArnott** [comment]
It turns out this is going to be a _huge_ job. I can't give a timeline for this yet.

**Comment by AArnott** [comment]
@siewers Just to throw it out there, since you did some great work over on Xunit.Combinatorial, if you're interested in doing more xunit v3 porting work, and want to expand to another xunit extension, let me know and we can collaborate on this one.

**Comment by siewers** [comment]
@AArnott, I'm happy to help, however this is an area that is very far from my comfort zone. I haven't worked with WPF or WinForms for a long time, but I might be able to help out.

**Comment by AArnott** [comment]
@siewers Nice. WPF/WInForms won't be the issue at all for this change, since that part already worked. And besides, there were only a few lines of code for each of those frameworks. 
Nearly all the code, and certainly all the changes we need to make for xunit v3 will be to adapt to the many breaking changes xunit v3 made to its own framework.

The work I've already done is in the xunitv3 branch. The first order of business is to get the compile errors resolved for the xunit.stafact project. I believe there are about 150 right now. Fixing many that remain will require understanding the new innerd's for v3, which I haven't done yet. I'm not sure how to best collaborate on this branch such that we avoid repeat work and merge conflicts though. Any suggestions? One way is to just take turns, I guess. I won't be working on this over the next few days at least, so if you were interested in making progress over the next few days, you'd be welcome to do just about anything.

**Comment by reduckted** [comment]
I haven't looked at exactly what's needed here, but I did [update some custom `Fact` attributes to work with xUnit v3](https://github.com/reduckted/GitWebLinks/commit/d526c3d725b4b88675c7bd963b47bd0bfa82df66) recently, so I'd be happy to help out if need be.

**Comment by AArnott** [comment]
I would welcome the help, @reduckted, thank you! Would you like to pick up where I left off as described above?

**Comment by AArnott** [comment]
Great. I've merged your changes thus far into my branch.

**Comment by reduckted** [comment]
@AArnott I've just pushed up more changes. Everything compiles now. The only problem is that some of the tests hang. 😢  I'm sure I've made a mistake somewhere in `UITestRunner.RunTest` with how the synchronization context is used, but I don't have the knowledge to find the problem.

**Comment by AArnott** [comment]
That sounds amazing, @reduckted. I'll check it out. I am much more knowledgeable when it comes to hangs than the xunit v3 api, so it sounds like you've done just enough. :)

**Comment by AArnott** [comment]
I've solved the hanging tests problem. 
The next step is to know why _some_ theories fail with missing test method arguments. The theories that work (e.g. `StaTheoryTests.ThreadAffinitizedDataObject`) go through `UIDelayEnumeratedTestCase`. The ones that fail (e.g. `StaTheoryTests.StaTheory_OnSTAThread`) go through `UITestCase`. 

@reduckted Do you think you can make headway on this? 

The theory template generates test cases with test arguments here:

https://github.com/AArnott/Xunit.StaFact/blob/1bbadb9a53d987457e07e3991c864916fb1fe7d3/src/Xunit.StaFact/Sdk/UITestCaseRunner.cs?plain=1#L44C9-L44C101

But it works or doesn't, depending on the type behind the `testCase` instance as I mention above. Yet the types and constructors for these two types appear to be built in equivalent ways.

**Comment by reduckted** [comment]
🤦‍♂ My bad. I was always passing null as the test method arguments instead of passing the given arguments. Fixed in this commit:
https://github.com/reduckted/Xunit.StaFact/commit/0495b9419288910a824be2258e2ec65287989852

I think the only tests that are still failing are `StaFactCtorSyncContextTests.SyncContextPreservedFromCtor`, The synchronization context that is set in the constructor is not preserved when the test is run. 🤔

**Comment by AArnott** [comment]
Great. I came up with nearly the same fix in parallel. But mine changed one other line which at the time was necessary to get tests to pass. Now, I find my fix is oddly not necessary for that, so I'm investigating a bit more.

**Comment by AArnott** [comment]
Well, the one line you added is in a slightly different place than one of mine, so I guess you killed two birds with one stone. I'm preparing a PR shortly. Thanks again!!

**Comment by reduckted** [comment]
@AArnott  Just an FYI. I've been using 2.0.29-alpha in a few different projects and everything is working smoothly. 🥳

**Comment by AArnott** [comment]
I'll be adopting it in a few projects this week too. If things go smoothly, I think I'll release a stable version.
